File: library-java/src/main/java/com/leeiidesu/loading/java/change/impl/CoverSwitchLayoutHelper.java
package com.leeiidesu.loading.java.change.impl;

import android.support.annotation.NonNull;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import com.leeiidesu.loading.java.change.SwitchLayoutHelper;

import java.security.acl.Group;

/**
 * Created by leeiides on 2017/6/23.
 */

class CoverSwitchLayoutHelper implements SwitchLayoutHelper {

    private View mOriginView;

    private ViewGroup mParentView;

    private int indexOfView;

    private View mCurrentView;

    private ViewGroup.LayoutParams originParams;

    private FrameLayout mFrameLayout;


    CoverSwitchLayoutHelper(@NonNull View view) {
        mOriginView = view;

        originParams = view.getLayoutParams();

        if (view.getParent() == null) {
            mParentView = (ViewGroup) view.getRootView().findViewById(android.R.id.content);
        } else {
            mParentView = (ViewGroup) view.getParent();
        }

        indexOfView = mParentView.indexOfChild(view);

        mCurrentView = view;

        mFrameLayout = new FrameLayout(view.getContext());
        mParentView.removeView(mCurrentView);
        mFrameLayout.addView(mCurrentView, originParams);

        mParentView.addView(mFrameLayout, indexOfView, originParams);
    }

    @Override
    public synchronized void switchLayout(@NonNull View targetView) {
        if (mCurrentView == targetView) return;

        if (mFrameLayout.getChildCount() == 2) {
            mFrameLayout.removeViewAt(1);
        }

        if (targetView != mOriginView) mFrameLayout.addView(targetView, 1, originParams);

        mCurrentView = targetView;

    }

    @Override
    public void removeAllViews() {
        mFrameLayout.removeView(mOriginView);
        mParentView.removeView(mFrameLayout);
        mParentView.addView(mOriginView, indexOfView, originParams);

        mFrameLayout = null;
        mParentView = null;
        mCurrentView = null;
        mOriginView = null;
        originParams = null;
    }

    @NonNull
    @Override
    public  View getCurrentStatusView() {
        return mCurrentView;
    }
}

File: library-kotlin/src/main/java/com/leeiidesu/loading/library/change/impl/CoverSwitchLayoutHelper.kt
package com.leeiidesu.loading.library.change.impl

import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.leeiidesu.loading.library.change.SwitchLayoutHelper

/**
 * 覆盖式切换
 * Created by leeiides on 2017/6/23.
 */
class CoverSwitchLayoutHelper(view: View) : SwitchLayoutHelper {

    private val mOriginView = view
    private val originParams: ViewGroup.LayoutParams = view.layoutParams

    private val mParentView: ViewGroup

    private var indexOfView: Int

    private var mCurrentView: View

    private var mFrameLayout: FrameLayout = FrameLayout(view.context)

    init {
        if (view.parent == null) {
            mParentView = view.rootView.findViewById(android.R.id.content) as ViewGroup
        } else {
            mParentView = view.parent as ViewGroup
        }

        indexOfView = mParentView.indexOfChild(view)

        mCurrentView = view

        mParentView.removeView(mCurrentView)
        mFrameLayout.addView(mCurrentView, originParams)

        mParentView.addView(mFrameLayout, indexOfView, originParams)
    }


    override fun switchLayout(targetView: View) {
        if (mCurrentView === targetView) return

        if (mFrameLayout.childCount == 2) {
            mFrameLayout.removeViewAt(1)
        }

        if (targetView !== mOriginView) mFrameLayout.addView(targetView, 1, originParams)

        mCurrentView = targetView
    }
}
